functional-samples/cookbook.sidepanel-open/service-worker.js

```diff
@@ -0,0 +1,44 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+chrome.runtime.onInstalled.addListener(() => {
+  chrome.contextMenus.create({
+    id: 'openSidePanel',
+    title: 'Open side panel',
+    contexts: ['all']
+  });
+  chrome.tabs.create({ url: 'page.html' });
+});
+
+chrome.contextMenus.onClicked.addListener((info, tab) => {
+  if (info.menuItemId === 'openSidePanel') {
+    // This will open the panel in all the pages on the current window.
+    chrome.sidePanel.open({ windowId: tab.windowId });
+  }
+});
+
+chrome.runtime.onMessage.addListener((message, sender) => {
+  // The callback for runtime.onMessage must return falsy if we're not sending a response
+  (async () => {
```

Comment:
Curious: why do we need the nested async function here?

Reply:
We can send a message response using the third callback argument, a sendResponse function.

Not every callback will send a response, and therefore we need to tell Chrome if we're going to send a response. Chrome uses the return value of the callback to determine whether to expect a response or not: if the return value is truthy, we have to call the sendResponse function.

This means we can't use an async function as our callback function because async functions return a Promise object (which is a truthy value, regardless of what value the Promise will contain). 

Therefore, if we're going to do something async but not send a message response, we need to find a way to return something like undefined.